### PR TITLE
fix: trade amount constants and confirm

### LINKED ISSUE #2962: MAX trades for BCH and DOGE giving "Not enough ETH to cover gas" error message
### Overview

When creating a BCH or DOGE trade, if you select the MAX amount, it gives the "Not enough ETH to cover gas" error message.

### References and additional details

1) Log into release.shapeshift.com and connect to native (Ensure Thorchain flag is on)
2) In the trade widget or on the Trade page create a BCH or DOGE trade and select the MAX button
3) Notice the "Not enough ETH to cover gas" error message

### Acceptance Criteria

The MAX amount for the send should not give error 

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### Estimated effort

_No response_

### Sponsor / Stakeholder

_No response_

### Bounty Hunters

- [Join our discord](https://discord.gg/shapeshift)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info
- Please ***do not*** start work on this issue until you are approved in Gitcoin.

### LINKED ISSUE #2979: THORChain switching output assets trade modal bug
### Overview

Switching outputs assets in a thorchain trade can make the modal jump back to the default pair page before going to the confirm trade page. 

### References and additional details

1. log into app.shapeshift.com with native 
2. turn on the thorswap flag 
3. go to the trade modal 
4. have ltc as input assets and fox as output
5. click preview trade button and go to the confirm page
6. go back to the pairing page
7. change the output asset to eth 
8. click preview trade button and notice the modal jumps back to the default pair (eth/fox) before moving on the the confirm page 

### Acceptance Criteria

trade modal should not jump to the default pair before moving to the confirm page 

### Need By Date

_No response_

### Screenshots/Mockups

https://user-images.githubusercontent.com/99363597/193928428-5529ccc9-2f73-467a-af89-415de2c854ff.mp4

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

draft - needs a well rested set of eyes before merging

this fixes a lot of stale value bugs we've been seeing by more aggressively guarding against invalid
inputs to the `getTradeAmountConstants` function

without doing this, it was possible to get into two very broken states

1) we should never be able to display a "Before Fees", or a "Protocol Fee" amount, without having
either a buy or sell amount, i.e. you can't possibly be receiving an amount "Before Fees" if you're
not swapping anything

2) we could see conditions where we were displaying a protocol fee for 0x quotes, or no protocol fee
for thorchain quotes. both of these are wrong.

it's possible we can revert the changes to `ReceiveSummary` and `TradeConfirm`, but both are
probably worthwhile

* receive summary should be purely presentational, and not calling `useGetTradeAmounts`, especiallya
  s it's used on the trade confirm screen
* it just seems prudent to snapshot the quote before we execute it, as there's no guarantees
  somethign can't mutate the values underneath us, and we display the wrong thing on the trade
  confirm screen

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #2979
closes #2962 

## Risk

* isolated to swapper v2, please test 0x and thor trades

### Engineering

* throw some logging after the guard conditions and before we call `getTradeAmountConstants` to
  ensure we always have necessary and sufficient data
* github actions is rekt, run CI locally before merging https://www.githubstatus.com/

### Operations

* **please always expand the "Expected Amount" section before previewing trades**
* if you are doing a thor trade, ensure you always see a protocol fee before previewing trades
* if you are doing a 0x trade, ensure you never see a protocol fee before previewing trades
* **please always expand the "Expected Amount" section before confirming trades**
* take a screenshot of the confirm trade page before confirming, and ensure amounts don't change
  after the transaction has confirmed on chain. watch the relevent block explorer of the sell asset
  for the source of truth.

## Screenshots (if applicable)

before - this should not be possible

Screen Shot 2022-10-05 at 12 13 20 AM

### CODE REVIEW

File: src/components/Trade/TradeConfirm/ReceiveSummary.tsx [hunk 1]
```diff
@@ -25,6 +24,7 @@ type ReceiveSummaryProps = {
   protocolFee?: string
   shapeShiftFee?: string
   slippage: number
+  swapperName: string
```
Comment (0xdef1cafe) [inline src/components/Trade/TradeConfirm/ReceiveSummary.tsx:27]: make dumb components dumber

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx [hunk 2]
```diff
@@ -161,6 +180,8 @@ export const TradeConfirm = ({ history }: RouterProps) => {
         return
       }
 
+      setExecutedTradeAmountConstants(tradeAmountConstants)
```
Comment (0xdef1cafe) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:183]: persist values used for the quote before executing the trade - display these with priority over values from form state which may change on us

File: src/components/Trade/TradeInputV2.tsx [hunk 3]
```diff
@@ -409,6 +411,7 @@ export const TradeInput = () => {
               protocolFee={tradeAmountConstants?.totalTradeFeeBuyAsset ?? ''}
               shapeShiftFee='0'
               slippage={slippage}
+              swapperName={swapperName}
```
Comment (0xdef1cafe) [inline src/components/Trade/TradeInputV2.tsx:414]: help make dum dum

File: src/components/Trade/hooks/useGetTradeAmounts.tsx [hunk 4]
```diff
@@ -196,21 +196,22 @@ export const useGetTradeAmounts = () => {
...
-          sellAssetUsdRate,
-          selectedCurrencyToUsdRate,
-          sellAssetTradeFeeUsd,
-          buyAssetTradeFeeUsd,
-          action,
-        })
-      : undefined
-
-  return tradeAmountConstants
+  if (!bnOrZero(buyTradeAsset?.amount).gt(0) || !bnOrZero(sellTradeAsset?.amount).gt(0)) return
```
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useGetTradeAmounts.tsx:199]: i'm pretty sure guarding against fees and amounts was the fix

File: src/components/Trade/hooks/useTradeAmounts.tsx [hunk 5]
```diff
@@ -233,6 +233,7 @@ export const useTradeAmounts = () => {
         setValue('sellAssetFiatRate', undefined)
         setValue('buyAssetFiatRate', undefined)
         setValue('feeAssetFiatRate', undefined)
+        setValue('fees', undefined)
```
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useTradeAmounts.tsx:236]: not sure if necessary

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx [hunk 6]
```diff
@@ -128,6 +136,17 @@ export const TradeConfirm = ({ history }: RouterProps) => {
...
+  useEffect(() => {
+    ;(async () => {
+      const buyAssetId = trade?.buyAsset.assetId
+      const sellAssetId = trade?.sellAsset.assetId
+      if (!buyAssetId || !sellAssetId) return ''
+      const swapperManager = await getSwapperManager(flags)
+      const bestSwapper = await swapperManager.getBestSwapper({ buyAssetId, sellAssetId })
+      setSwapperName(bestSwapper?.name ?? '')
+    })()
+  }, [flags, trade])
```
Comment (gomesalexandre) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:148]: Wondering if we want to be extra safu here and have 0 deps:
- changing flags would re-render swapper
- `trade` should always be defined, and changing it with user input means we unmounted `<TradeConfirm />` and re-mounted it

Looks like there's no case where we legitimately *need* the deps, but there's room for bugs by having them

**Review by gomesalexandre — APPROVED** [review]
Paired on this one, confirmed working for 0x and Thor swaps - before fees, swapper name and fees are now stable at Confirm step

**Comment by 0xdef1cafe** [comment]
merging as CI passed locally and github actions are degraded

**Pull request merged**